### ISSUE TITLE
IHost.RunAsync() never completes

### ISSUE THREAD

**Issue by martyt** [dotnet/runtime#44086]
### Describe the bug
I'm building a .NET Core 3.1 app that will run a `BackgroundService` in a Docker container. While I've implemented the startup and shutdown tasks for the BackgroundService and the service is definitely shutting down when triggered via `SIGTERM`, I'm finding that the `await host.RunAsync()` call never completes - meaning the remaining code in my `Main()` block isn't executed.

### To Reproduce
Use latest Microsoft.Extensions.Hosting package 

Program.cs:
```
using System;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Hosting;

namespace BackgroundServiceTest
{
    class Program
    {
        static async Task Main(string[] args)
        {
            Console.WriteLine("Main: starting");
            try
            {
                using var host = CreateHostBuilder(args).Build();

                Console.WriteLine("Main: Waiting for RunAsync to complete");

                await host.RunAsync();

                Console.WriteLine("Main: RunAsync has completed");
            }
            finally
            {
                Console.WriteLine("Main: stopping");
            }
        }

        public static IHostBuilder CreateHostBuilder(string[] args) =>
            Host.CreateDefaultBuilder(args)
                .UseConsoleLifetime()
                .ConfigureServices((hostContext, services) =>
                {
                    services.AddHostedService<Worker>();

                    // give the service 120 seconds to shut down gracefully before whacking it forcefully
                    services.Configure<HostOptions>(options => options.ShutdownTimeout = TimeSpan.FromSeconds(120));
                });

    }

    class Worker : BackgroundService
    {
        protected override async Task ExecuteAsync(CancellationToken stoppingToken)
        {
            Console.WriteLine("Worker: ExecuteAsync called...");
            try
            {
                while (!stoppingToken.IsCancellationRequested)
                {
                    await Task.Delay(TimeSpan.FromSeconds(10), stoppingToken);
                    Console.WriteLine("Worker: ExecuteAsync is still running...");
                }
            }
            catch (OperationCanceledException) // will get thrown if TaskDelay() gets cancelled by stoppingToken
            {
                Console.WriteLine("Worker: OperationCanceledException caught...");
            }
            finally
            {
                Console.WriteLine("Worker: ExecuteAsync is terminating...");
            }
        }

        public override Task StartAsync(CancellationToken cancellationToken)
        {
            Console.WriteLine("Worker: StartAsync called...");
            return base.StartAsync(cancellationToken);
        }

        public override async Task StopAsync(CancellationToken cancellationToken)
        {
            Console.WriteLine("Worker: StopAsync called...");
            await base.StopAsync(cancellationToken);
        }

        public override void Dispose()
        {
            Console.WriteLine("Worker: Dispose called...");
            base.Dispose();
        }
    }
}
```
Dockerfile: 
```
#See https://aka.ms/containerfastmode to understand how Visual Studio uses this Dockerfile to build your images for faster debugging.

FROM mcr.microsoft.com/dotnet/core/runtime:3.1-buster-slim AS base
WORKDIR /app

FROM mcr.microsoft.com/dotnet/core/sdk:3.1-buster AS build
WORKDIR /src
COPY ["BackgroundServiceTest.csproj", "./"]
RUN dotnet restore "BackgroundServiceTest.csproj"
COPY . .
WORKDIR "/src/"
RUN dotnet build "BackgroundServiceTest.csproj" -c Release -o /app/build

FROM build AS publish
RUN dotnet publish "BackgroundServiceTest.csproj" -c Release -o /app/publish

FROM base AS final
WORKDIR /app
COPY --from=publish /app/publish .
ENTRYPOINT ["dotnet", "BackgroundServiceTest.dll"]
```

docker-compose.yml:
```
version: '3.4'

services:
  backgroundservicetest:
    image: ${DOCKER_REGISTRY-}backgroundservicetest
    build:
      context: .
      dockerfile: Dockerfile
```

Start the app with `docker-compose up --build` and see the console logging from the application indicating that the service is running. 

Run a second command prompt window and execute `docker stop -t 90 backgroundservicetest_backgroundservicetest_1` and see that the app shuts down gracefully but never writes any of the console messages after the call to `await host.RunAsync()`

### Expected behavior
After the BackgroundService completes shutting down, control should be returned to the method that called `.RunAsync()`

### Screenshots
Console output from docker-compose 
```
    Successfully built 3aa605d4798f
    Successfully tagged backgroundservicetest:latest
    Recreating backgroundservicetest_backgroundservicetest_1 ... done
    Attaching to backgroundservicetest_backgroundservicetest_1
    backgroundservicetest_1  | Main: starting
    backgroundservicetest_1  | Main: Waiting for RunAsync to complete
    backgroundservicetest_1  | Worker: StartAsync called...
    backgroundservicetest_1  | Worker: ExecuteAsync called...
    backgroundservicetest_1  | info: Microsoft.Hosting.Lifetime[0]
    backgroundservicetest_1  |       Application started. Press Ctrl+C to shut down.
    backgroundservicetest_1  | info: Microsoft.Hosting.Lifetime[0]
    backgroundservicetest_1  |       Hosting environment: Production
    backgroundservicetest_1  | info: Microsoft.Hosting.Lifetime[0]
    backgroundservicetest_1  |       Content root path: /app
    backgroundservicetest_1  | Worker: ExecuteAsync is still running...
    backgroundservicetest_1  | Worker: ExecuteAsync is still running...
    backgroundservicetest_1  | info: Microsoft.Hosting.Lifetime[0]
    backgroundservicetest_1  |       Application is shutting down...
    backgroundservicetest_1  | Worker: StopAsync called...
    backgroundservicetest_1  | Worker: OperationCanceledException caught...
    backgroundservicetest_1  | Worker: ExecuteAsync is terminating...
    backgroundservicetest_1  | Worker: Dispose called...
    backgroundservicetest_backgroundservicetest_1 exited with code 0
```

**Comment by pinkfloydx33** [comment]
I know this sound silly, but have you tried using a `using` *block* instead? I seem to recall a similar issue that I had which was resolved that way. 

```csharp
var host = CreateHostBuilder(args).Build();
using(host) {
   await host.RunAsync();
} 
Console.WriteLine("...");
```

**Comment by maryamariyan** [comment]
`e.Cancel` in:

https://github.com/dotnet/runtime/blob/857fe5b3f079b59d45cdf3d3a2050c35c4a9d370/src/libraries/Microsoft.Extensions.Hosting/src/Internal/ConsoleLifetime.cs#L92-L96

is supposed to make sure we don't terminate the process immediately, and rather wait for the main thread to exit gracefully.

> the await host.RunAsync() call never completes - meaning the remaining code in my Main() block isn't executed

I think termination via SIGTERM might be resulting in immediate exit from the process since perhaps `ConsoleCancelEventArgs.Cancel` is not supported on non-windows platforms. 

---

Also refer to [documentation](https://docs.microsoft.com/en-us/dotnet/api/system.consolecanceleventargs.cancel?view=net-5.0): `e.Cancel` is set to false by default, causing the current process to terminate when false and when set to true, the current process should resume when the event handler concludes.

**Comment by maryamariyan** [comment]
@martyt using Ctrl+C to terminate, (1)on windows, (2) on ubuntu 20.4, and (3) on docker container (using linux centos 7), results in the same output for the repro code you provided in the issue description.

> the service is definitely shutting down when triggered via SIGTERM

are you not using ctrl+c to terminate?

```
Main: starting
Main: Waiting for RunAsync to complete
Worker: StartAsync called...
Worker: ExecuteAsync called...
info: Microsoft.Hosting.Lifetime[0]
      Application started. Press Ctrl+C to shut down.
info: Microsoft.Hosting.Lifetime[0]
      Hosting environment: Production
info: Microsoft.Hosting.Lifetime[0]
      Content root path: /HostRepro
Worker: ExecuteAsync is still running...
^Cinfo: Microsoft.Hosting.Lifetime[0]
      Application is shutting down...
Worker: StopAsync called...
Worker: OperationCanceledException caught...
Worker: ExecuteAsync is terminating...
Worker: Dispose called...
Main: RunAsync has completed
Main: stopping
```

cc: @tarekgh

**Comment by martyt** [comment]
@maryamariyan No, I'm not using Ctrl+C - I'm relying on `SIGTERM` being sent to the service when Docker is shutting down the container (as happens when executing `docker stop -t 90 backgroundservicetest_backgroundservicetest_1`). 

Just ran my repro again after updating `Microsoft.Extensions.Hosting` to 3.1.13 and using the latest `runtime:3.1-buster-slim` docker image. Same results as originally reported.

**Comment by tarekgh** [comment]
@martyt if the docker container is shutting down anyway, why you care having the AsuncRun to exit? I don't think you need to depend on that.

**Comment by maryamariyan** [comment]
I see. I think there is no support for using `SIGTERM` for shutting down due to the explanation in https://github.com/dotnet/runtime/issues/44086#issuecomment-809874801

**Comment by davidfowl** [comment]
We support SIGTERM out of the box (hooking process exit).

**Comment by maryamariyan** [comment]
@martyt do you mind please testing this on 5.0? @ericstj mentioned there was a runtime fix made for this on the last release.

**Comment by ericstj** [comment]
Right, this looks like a duplicate of https://github.com/dotnet/runtime/issues/36089 to me.  That was fixing ProcessExit on docker on Windows.

**Comment by martyt** [comment]
> @martyt if the docker container is shutting down anyway, why you care having the AsuncRun to exit? I don't think you need to depend on that.

@tarekgh - If `.RunAsync()` never returns, it's difficult to do any required cleanup prior to application exit. Doesn't matter what I need to run - the bottom line is, I have code after `.RunAsync()` that isn't getting executed.

**Comment by davidfowl** [comment]
Unfortunately this is by design, at least currently without other bigger changes to the runtime. When the process is being torn down we block the event handler until the host is dispose. RunAsync and Run automatically unblock the waiting so code will not execute after it (https://github.com/dotnet/runtime/blob/82ca681cbac89d813a3ce397e0c665e6c051ed67/src/libraries/Microsoft.Extensions.Hosting.Abstractions/src/HostingAbstractionsHostExtensions.cs#L75). That's what disposing the host does.

@ericstj That issue you mentioned is similar but windows specific. With this one, SIGTERM does trigger process exit but once the host is disposed, it will unwind.

@martyt 

To work around this you'd need to write code like this:

```C#
static async Task Main(string[] args)
{
    Console.WriteLine("Main: starting");
    IHost host = null;
    try
    {
        host = CreateHostBuilder(args).Build();

        Console.WriteLine("Main: Waiting for RunAsync to complete");

        await host.WaitForShutdownAsync();

        Console.WriteLine("Main: RunAsync has completed");
    }
    finally
    {
        Console.WriteLine("Main: stopping");

        if (host is IAsyncDisposable d) await d.DisposeAsync();
    }
}
```

**Comment by martyt** [comment]
@davidfowl Thanks for the suggested workaround, but it looks like the call to `host.RunAsync()` is missing. 

So I added that call (without `await`ing, which makes me feel dirty...) and then after adding the `await host.WaitForShutdownAsync()`, the app performed as I would expect -- control returned to `Main()` and all of the shutdown / cleanup code was executed:

```
backgroundservicetest_1  | Worker: ExecuteAsync is still running...
backgroundservicetest_1  | info: Microsoft.Hosting.Lifetime[0]
backgroundservicetest_1  |       Application is shutting down...
backgroundservicetest_1  | Worker: StopAsync called...
backgroundservicetest_1  | Worker: StopAsync called...
backgroundservicetest_1  | Worker: OperationCanceledException caught...
backgroundservicetest_1  | Worker: ExecuteAsync is terminating...
backgroundservicetest_1  | Main: RunAsync has completed
backgroundservicetest_1  | Main: stopping
backgroundservicetest_1  | Worker: Dispose called...
backgroundservicetest_backgroundservicetest_1 exited with code 0
```

**Comment by davidfowl** [comment]
Ah I forgot to call StartAsync in my sample. You don't want to call RunAsync because it could dispose without you observing it explicitly. With the pattern below you control disposal.

```C#
static async Task Main(string[] args)
{
    Console.WriteLine("Main: starting");
    IHost host = null;
    try
    {
        host = CreateHostBuilder(args).Build();

        Console.WriteLine("Main: Waiting for RunAsync to complete");
        await host.StartAsync();

        await host.WaitForShutdownAsync();

        Console.WriteLine("Main: RunAsync has completed");
    }
    finally
    {
        Console.WriteLine("Main: stopping");

        if (host is IAsyncDisposable d) await d.DisposeAsync();
    }
}
```

**Comment by martyt** [comment]
@davidfowl Thanks, that seems like a perfectly reasonable workaround. (I still think `.RunAsync()` should return control to the caller, but that's less important to me than having a way to achieve the result that I need...)

**Comment by davidfowl** [comment]
Agreed. Closing this in favor of https://github.com/dotnet/runtime/issues/50527
